metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ESGraphQueryDAO.java

```diff
@@ -0,0 +1,163 @@
+      @Nonnull  final Filter sourceEntityFilter,
+      @Nullable final String destinationType,
+      @Nonnull final Filter destinationEntityFilter,
+      @Nonnull final List<String> relationshipTypes,
+      @Nonnull final RelationshipFilter relationshipFilter,
+      final int offset,
+      final int count) {
+    // also delete any relationship going to or from it
+    final RelationshipDirection relationshipDirection = relationshipFilter.getDirection();
+
+    SearchRequest searchRequest = new SearchRequest();
+
+    SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+
+    searchSourceBuilder.from(offset);
+    searchSourceBuilder.size(count);
+
+    BoolQueryBuilder finalQuery = QueryBuilders.boolQuery();
+
+    // set source filter
+    String sourceNode = relationshipDirection == RelationshipDirection.OUTGOING ? "source" : "destination";
+    if (sourceType != null && sourceType.length() > 0) {
+      finalQuery.must(QueryBuilders.termQuery(sourceNode + ".entityType", sourceType));
+    }
+    addCriterionToQueryBuilder(sourceEntityFilter.getCriteria(), sourceNode, finalQuery);
+
+    // set destination filter
+    String destinationNode = relationshipDirection == RelationshipDirection.OUTGOING ? "destination" : "source";
+    if (destinationType != null && destinationType.length() > 0) {
+      finalQuery.must(QueryBuilders.termQuery(destinationNode + ".entityType", destinationType));
+    }
+    addCriterionToQueryBuilder(destinationEntityFilter.getCriteria(), destinationNode, finalQuery);
+
+    // set relationship filter
+    if (relationshipTypes.size() > 0) {
+      BoolQueryBuilder relationshipQuery = QueryBuilders.boolQuery();
+      relationshipTypes.forEach(relationshipType
+          -> relationshipQuery.should(QueryBuilders.termQuery("relationshipType", relationshipType)));
+      finalQuery.must(relationshipQuery);
+    }
+
+    searchSourceBuilder.query(finalQuery);
+
+    searchRequest.source(searchSourceBuilder);
+
+    searchRequest.indices(indexConvention.getIndexName(INDEX_NAME));
+
+    try {
+      final SearchResponse searchResponse = client.search(searchRequest, RequestOptions.DEFAULT);
+      return searchResponse;
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return null;
+  }
+
+  public BulkByScrollResponse deleteByQuery(
+      @Nullable final String sourceType,
+      @Nonnull  final Filter sourceEntityFilter,
+      @Nullable final String destinationType,
+      @Nonnull final Filter destinationEntityFilter,
+      @Nonnull final List<String> relationshipTypes,
+      @Nonnull final RelationshipFilter relationshipFilter) {
+    // also delete any relationship going to or from it
+    final RelationshipDirection relationshipDirection = relationshipFilter.getDirection();
+
+    DeleteByQueryRequest deleteByQueryRequest = new DeleteByQueryRequest();
+
+    BoolQueryBuilder finalQuery = QueryBuilders.boolQuery();
+
+    // set source filter
```

Comment:
Can we share this source / dest filter building logic?

Reply:
updated!